MB-12926 add weight tickets table and model

felipe-lee opened the pull request:
## [Jira ticket](https://dp3.atlassian.net/browse/MB-12926) for this change

The ticket above is to track the work, but the details are in the [discovery ticket](https://dp3.atlassian.net/browse/MB-12857).

## Summary

This adds a new `WeightTicket` model and a corresponding `weight_tickets` table that will be used to track weight tickets for PPMs.

Do the field names and comments on the table columns make sense?

## Checking Changes

<details>
<summary>💻 Terminal setup</summary>

### Run migrations on top of existing db

You can run this the first time you pull it down to see it applies the migration properly on top of the existing DB. Not required though, you can also run the other command instead.

```sh
make db_dev_migrate
```

### Run all migrations

```sh
make db_dev_fresh
```

</details>

### Additional steps 

1. Access the DB however you usually do and see if the new table is there correctly

## Verification Steps for Author

These are to be checked by the author.

- [x] Have the Jira acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Database

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/dev/contributing/database/Database-Migrations#zero-downtime-migrations)
- [ ] Have been communicated to #g-database
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/dev/contributing/database/Database-Migrations#secure-migrations)

robot-mymove commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
  DangerID: danger-id-Danger;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:link: <a href="https://dp3.atlassian.net/browse/MB-12926">MB-12926</a></td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 6066515c8a2635357045797046d39672291b4e9a
</p>

ronaktruss commented on migrations/app/schema/20220622225529_create_weight_tickets_table.up.sql:
```diff
@@ -0,0 +1,37 @@
+COMMENT on COLUMN weight_tickets.full_weight IS 'Stores the weight of the vehicle when full.';
+COMMENT on COLUMN weight_tickets.has_full_weight_ticket IS 'Indicates if the customer has a weight ticket for the vehicle weight when full.';
+COMMENT on COLUMN weight_tickets.owns_trailer IS 'Indicates if the customer used a trailer they own for the move.';
+COMMENT on COLUMN weight_tickets.trailer_meets_criteria IS 'Indicates if the trailer that the customer used meets all the criteria to be claimable.';
```
I'm not sure how much value there is for adding comments to for the columns that store uuids for other tables (foreign keys), but it looks like our other migrations have comments for them.

felipe-lee replied:
Ah, I think like Reggie said below, it looks like we haven't done it for recent tables so I'd left it out: https://github.com/transcom/mymove/pull/8809/files#r905062840

I can add them though since we do have 3 links to the same table.

ronaktruss commented on pkg/models/weight_ticket.go:
```diff
@@ -0,0 +1,33 @@
+	VehicleDescription                *string     `json:"vehicle_description" db:"vehicle_description"`
+	EmptyWeight                       *unit.Pound `json:"empty_weight" db:"empty_weight"`
+	HasEmptyWeightTicket              *bool       `json:"has_empty_weight_ticket" db:"has_empty_weight_ticket"`
+	EmptyDocumentID                   *uuid.UUID  `json:"empty_document_id" db:"empty_document_id"`
+	EmptyDocument                     *Document   `belongs_to:"documents" fk_id:"empty_document_id"`
```
Since we are planning on preemptively creating the documents do we still want to have these as pointers?

felipe-lee replied:
I kept them as pointers in case we end up pivoting to variation 1 of our plans. I figured this worked for both variations so this might be easier. I can update them though if you'd prefer them to not be pointers since like you said, our current plan is to create them all together.

ronaktruss replied:
That's fine to keep them as pointers then, but it might we worth a comment indicating that we expect these to exist

felipe-lee replied:
Sure, I can add that

felipe-lee replied:
I added it to the comment on the type since these rows are already pretty long, but I can move it if you think it's better to be right next to the field. I guess I could also add them above the field...which i didn't think of till just now 🤦🏼

felipe-lee replied:
Ah, actually putting them on their own lines messes up the go formatting. So, do you prefer them to be on the same line and it just be long, or above with the `WeghtTicket` comment (lines 11-13)

ronaktruss replied:
Personally, I'd probably put it **above** the lines and causing the alignment issues because imo it makes the comment more visible. Feel free to ignore my personal preference though

felipe-lee replied:
Based on our conversation in slack, these are no longer pointers so I've removed the comments about that.

reggieriser commented on migrations/app/schema/20220622225529_create_weight_tickets_table.up.sql:
```diff
@@ -0,0 +1,37 @@
+COMMENT on COLUMN weight_tickets.full_weight IS 'Stores the weight of the vehicle when full.';
+COMMENT on COLUMN weight_tickets.has_full_weight_ticket IS 'Indicates if the customer has a weight ticket for the vehicle weight when full.';
+COMMENT on COLUMN weight_tickets.owns_trailer IS 'Indicates if the customer used a trailer they own for the move.';
+COMMENT on COLUMN weight_tickets.trailer_meets_criteria IS 'Indicates if the trailer that the customer used meets all the criteria to be claimable.';
```
**thought:** I thought we were putting comments on all columns (pretty sure that's what we did when we first commented the database).  But I do see that some recent table creation doesn't have them on primary/foreign keys or timestamps (presumably because those are more self-explanatory).

felipe-lee replied:
I think that's why I'd left them out, since recent ones didn't have them. I'll add them for this one though since we have three fields pointing at the same table

felipe-lee replied:
Added now!

reggieriser reviewed (commented):
The migration ran fine and generally looks good.  Had a few smallish items for your consideration, though.

felipe-lee commented on migrations/app/schema/20220622225529_create_weight_tickets_table.up.sql:
```diff
@@ -0,0 +1,37 @@
+CREATE TABLE weight_tickets
+(
+	id uuid PRIMARY KEY NOT NULL,
```
Ah cool, I can drop it

felipe-lee commented on migrations/app/schema/20220622225529_create_weight_tickets_table.up.sql:
```diff
@@ -0,0 +1,37 @@
+	ppm_shipment_id uuid NOT NULL
+		CONSTRAINT weight_tickets_ppm_shipments_id_fkey
+		REFERENCES ppm_shipments,
+	created_at timestamp NOT NULL,
+	updated_at timestamp NOT NULL,
```
I think we do, and for the sql I don't mind doing that. 

I mainly did these in the same order as the model and I didn't put them at the end of the model because fields end up getting put at the ends of models sometimes which then leaves these fields in the middle somewhere rather than at the end. So I figure closer to the beginning in a more static spot is better maybe. I can update this

felipe-lee commented on migrations/app/schema/20220622225529_create_weight_tickets_table.up.sql:
```diff
@@ -0,0 +1,37 @@
+		REFERENCES ppm_shipments,
+	created_at timestamp NOT NULL,
+	updated_at timestamp NOT NULL,
+	deleted_at timestamptz,
```
I think this came up from this conversation on the two: https://ustcdp3.slack.com/archives/CSGDM3NUW/p1653513178071059

So I think we're using the `timestamp` for the `created_at` and `updated_at` since that's what `Pop` expects, but those don't have tz data while the `timestampz` does. Do you think we should just use the plain `timestamp` for all fields for consistency?

reggieriser replied:
Well, I just looked at most `deleted_at` fields appear to use timezones like this (but not all), so this is probably fine.  It'd be nice (not for this PR) to have more consistency and/or guidelines for timestamps.  I recall briefly looking into this at one point and thinking it didn't have a practical difference since we store timestamps in UTC I believe.

felipe-lee commented on migrations/app/schema/20220622225529_create_weight_tickets_table.up.sql:
```diff
@@ -0,0 +1,37 @@
+	proof_of_trailer_ownership_document_id uuid
+		CONSTRAINT weight_tickets_proof_of_trailer_ownership_document_id_fkey
+		REFERENCES documents
+);
```
Yeah, we're going to be creating these pretty barebones because of the way uploads for this model will work. The users will likely be uploading docs before they fill in the data.

felipe-lee replied:
Based on conversations around validation in here and on slack, there are a few more required fields now (the empty and full document id fields).

felipe-lee commented on migrations/app/schema/20220622225529_create_weight_tickets_table.up.sql:
```diff
@@ -0,0 +1,37 @@
+		REFERENCES documents
+);
+
+CREATE INDEX weight_tickets_deleted_at_idx ON weight_tickets USING btree (deleted_at);
```
I don't know if we'd be doing searches by the foreign keys on this table. I'd be fine with adding them, but don't know if we should wait like you said since it might be premature. 

@ronaktruss you've at the weight docs more, what do you think?

felipe-lee replied:
Oh right, Idk why I was more focused on the documents. I'll add that index!

felipe-lee commented on pkg/models/weight_ticket.go:
```diff
@@ -0,0 +1,33 @@
+	"github.com/transcom/mymove/pkg/unit"
+)
+
+type WeightTicket struct {
```
I didn't get an error, but am happy to add a comment. Probably for the best to make sure people know exactly what it's for. Or at least what it was meant for when we added it 😆

felipe-lee commented on pkg/models/weight_ticket.go:
```diff
@@ -0,0 +1,33 @@
+	ProofOfTrailerOwnershipDocument   *Document   `belongs_to:"documents" fk_id:"proof_of_trailer_ownership_document_id"`
+}
+
+type WeightTickets []WeightTicket
```
We have been doing required field checks in the serv obj validation. 

For the enums, since we use constants and go is strongly typed, isn't that covered already, or can those fields actually be set to something else? Like I know the base type of our constants are usually regular types (e.g. strings), but I thought that since we set the field to be of that unique type, only those values (and whatever the empty value is). Or did you mean something else?

Hmm I agree it's possible we might have interactions with the db that doesn't go through service objects (e.g. our testdatagen `Make<model>` functions).  I wonder if we need to add better documentation around this then, especially to disambiguate which validation goes where.

reggieriser replied:
>For the enums, since we use constants and go is strongly typed, isn't that covered already, or can those fields actually be set to something else?

Here's a [Go playground](https://go.dev/play/p/xXxJ-3yDW2Q) with an example of what I was thinking about in regards to string "enums" in Go.  That said, if the database has a proper enum defined, then it should catch any invalid values (albeit as a potentially user-unfriendly database error).  Enums might also be detected at the swagger level if defined there.  But I don't think Go by itself will stop other string literals outside the constants from being used, at least the way we have those usually defined.

felipe-lee replied:
Ah, thank you for sharing that playground example. Hmm yeah, I feel like this is one of those that would probably be good to have at the model validation level.

We do have certain checks sometimes at the service object level for this, but usually it's more checking if it's a specific value to then do other validation, not checking that it is a valid value. We do have them defined at the swagger level for the ones I've seen, but that's not a great thing to rely on probably.

felipe-lee replied:
I've added some model-level validation now

duncan-truss commented on pkg/models/ppm_shipment.go:
```diff
@@ -77,6 +77,7 @@ type PPMShipment struct {
 	SITEstimatedEntryDate          *time.Time        `json:"sit_estimated_entry_date" db:"sit_estimated_entry_date"`
 	SITEstimatedDepartureDate      *time.Time        `json:"sit_estimated_departure_date" db:"sit_estimated_departure_date"`
 	SITEstimatedCost               *unit.Cents       `json:"sit_estimated_cost" db:"sit_estimated_cost"`
+	WeightTickets                  WeightTickets     `has_many:"weight_tickets" fk_id:"ppm_shipment_id" order_by:"created_at desc"`
```
I was thinking because we need to number trip numbers (Trip 1, Trip 2, ...) that we could return the weight tickets in ascending order of earliest creation date.

felipe-lee replied:
Updated this to now be `asc` instead of `desc`

felipe-lee commented on migrations/app/schema/20220622225529_create_weight_tickets_table.up.sql:
```diff
@@ -0,0 +1,42 @@
+	deleted_at timestamptz
+);
+
+CREATE INDEX weight_tickets_ppm_shipment_id_idx ON weight_tickets USING hash (ppm_shipment_id);
```
@reggieriser @ronaktruss Here's the index on the ppm shipment ID. I used `hash` since we won't be checking for `null` and afaik will only be matching IDs exactly. Lmk if you think it should be `btree` though (or something else). 
Based my decision on the info here: https://www.postgresqltutorial.com/postgresql-indexes/postgresql-index-types/

felipe-lee commented:
@reggieriser @ronaktruss @duncan-truss  I think I've addressed all the comments, but lmk if I missed anything, or if anything else needs to be addressed in the new stuff.

duncan-truss commented on migrations/app/schema/20220622225529_create_weight_tickets_table.up.sql:
```diff
@@ -0,0 +1,42 @@
+		REFERENCES ppm_shipments,
+	vehicle_description varchar,
+	empty_weight int,
+	has_empty_weight_ticket bool,
```
This isn't a blocker but the phrasing on the frontend is "I don't have this weight ticket" so it might be easier to frame this column where the default is the unchecked false value to match.  That's kind of why I went with missingEmptyWeightTicket so we don't have to invert the boolean.  Ditto for missing full weight.
